File: app/models/reservation/hourly_price_calculator.rb
class Reservation::HourlyPriceCalculator
  def initialize(order)
    @order = order
    @pricing = @order.transactable_pricing
  end

  def price
    # Price is for each day, hours reserved in day * hourly price
    @order.periods.map do |period|
      price_for_hours(period.hours)
    end.sum.to_money
  end

  def valid?
    !@order.periods.empty? && @order.periods.all? { |p| p.hours > 0 && @order.minimum_booking_minutes <= p.minutes }
  end

  def price_for_hours(hours)
    prices = @pricing.try(:all_prices_for_unit)
    if prices
      block_size = prices.keys.sort.inject do |largest_block, block_days|
        largest_block = block_days if hours >= block_days
        largest_block
      end
      pricing = prices[block_size]
      price = pricing[:price]

      if @order.favourable_pricing_rate || hours < block_size
        (((hours / block_size.to_f) * price.cents).round / BigDecimal.new(price.currency.subunit_to_unit)).to_money(price.currency)
      else
        priced_hours = hours / block_size
        left_hours = hours - priced_hours * block_size
        calculated_price = ((priced_hours * price.cents).round / BigDecimal.new(price.currency.subunit_to_unit)).to_money(price.currency)
        if left_hours.zero?
          calculated_price
        else
          calculated_price + price_for_days(left_hours)
        end
      end
    end
  end

  private

  def listing
    @order.transactable
  end
end

File: app/frontend/javascripts/sections/bookings/price_calculator/hourly_price_calculator.js
var AdditionalChargesCalculator, HourlyPriceCalculator;

AdditionalChargesCalculator = require('../additional_charges');

HourlyPriceCalculator = function() {
  function HourlyPriceCalculator(listing) {
    this.listing = listing;
    this.additionalCharges = new AdditionalChargesCalculator(
      $('#additional-charges-' + this.listing.id),
      this.listing.data.subunit_to_unit_rate
    );
  }

  HourlyPriceCalculator.prototype.getPrice = function() {
    var total;
    total = this.priceForHours(this.listing.minutesBooked() / 60) * this.listing.getQuantity();
    total += this.additionalCharges.getCharges(total);
    if (this.pricing) {
      this.listing.currentPricingId = this.pricing.id;
    }
    return total;
  };

  HourlyPriceCalculator.prototype.priceForHours = function(hours) {
    var block_size, calculated_price, left_hours, price, priced_hours, prices, pricesHours;
    prices = this.listing.pricesByHours;
    pricesHours = _.keys(prices);
    if (pricesHours.length === 0) {
      return 0;
    }
    block_size = _.inject(pricesHours, function(largestBlock, blockDays) {
      if (hours >= blockDays) {
        largestBlock = blockDays;
      }
      return largestBlock;
    });
    this.pricing = prices[block_size];
    price = this.pricing.price;
    if (this.listing.hasFavourablePricingRate() || hours < block_size) {
      return Math.round(hours / block_size * price);
    } else {
      priced_hours = Math.floor(hours / block_size);
      left_hours = hours - priced_hours * block_size;
      calculated_price = Math.round(priced_hours * price);
      if (left_hours === 0) {
        return calculated_price;
      } else {
        return calculated_price + this.priceForHours(left_hours);
      }
    }
  };

  return HourlyPriceCalculator;
}();

module.exports = HourlyPriceCalculator;
